src/resource-rental/resource-rental.ts

```diff
@@ -129,29 +129,43 @@ export class ResourceRental {
         abortController.abort(signalOrTimeout.reason);
       }
     }
-
-    const activity = await this.activityModule.createActivity(this.agreement);
-    this.currentExeUnit = await this.activityModule.createExeUnit(activity, {
-      storageProvider: this.storageProvider,
-      networkNode: this.resourceRentalOptions?.networkNode,
-      executionOptions: this.resourceRentalOptions?.activity,
-      signalOrTimeout: abortController.signal,
-      ...this.resourceRentalOptions?.exeUnit,
-    });
-
-    return this.currentExeUnit;
+    return this.createExeUnit(abortController.signal);
   }
 
+  /**
+   * Destroy previously created exe-unit.
+   * Please note that if ResourceRental is left without ExeUnit for some time (default 90s)
+   * the provider will terminate the Agreement and ResourceRental will be unuseble
+   */
   async destroyExeUnit() {
-    if (this.currentExeUnit) {
+    if (this.currentExeUnit !== null) {
       await this.activityModule.destroyActivity(this.currentExeUnit.activity);
       this.currentExeUnit = null;
     } else {
-      throw new Error(`There is no activity to destroy.`);
+      throw new GolemUserError(`There is no exe-unit to destroy.`);
     }
   }
 
   async fetchAgreementState() {
     return this.marketModule.fetchAgreement(this.agreement.id).then((agreement) => agreement.getState());
   }
+
+  private async createExeUnit(abortSignal: AbortSignal) {
+    if (!this.exeUnitPromise) {
+      this.exeUnitPromise = (async () => {
+        const activity = await this.activityModule.createActivity(this.agreement);
+        this.currentExeUnit = await this.activityModule.createExeUnit(activity, {
+          storageProvider: this.storageProvider,
+          networkNode: this.resourceRentalOptions?.networkNode,
+          executionOptions: this.resourceRentalOptions?.activity,
+          signalOrTimeout: abortSignal,
+          ...this.resourceRentalOptions?.exeUnit,
+        });
+        return this.currentExeUnit;
+      })().finally(() => {
```

Comment:
Please provide basic error handling. `createActivity` or `createExeUnit` can throw and this code silences the error.